server/src/main/java/org/elasticsearch/index/engine/ThreadPoolMergeExecutorService.java

```diff
@@ -53,9 +154,9 @@ public class ThreadPoolMergeExecutorService {
      * The merge tasks that are waiting execution. This does NOT include backlogged or currently executing merge tasks.
      * For instance, this can be empty while there are backlogged merge tasks awaiting re-enqueuing.
      */
-    private final PriorityBlockingQueue<MergeTask> queuedMergeTasks = new PriorityBlockingQueue<>(
-        64,
-        Comparator.comparingLong(MergeTask::estimatedMergeSize)
+    private final PriorityBlockingQueueWithMaxLimit<MergeTask> queuedMergeTasks = new PriorityBlockingQueueWithMaxLimit<>(
+        MergeTask::estimatedMergeSize,
+        Long.MAX_VALUE
```

Comment:
If we cannot read the available disk space, all merging will be allowed (and there is some logging complaining about it).

Reply:
Similarly, merging starts without waiting to read the fs stats.

Generally, my thinking was to let the merging execute as before, and if the filesystem stats are available, and if the remaining disk space is low, only then stop new merges from starting.

Alternatively, we could allow merging to happen only after filesystem stats are available and the remaining disk space is sufficient.

WDYT?